contracts/BookMarketplace.sol

```diff
@@ -290,10 +309,76 @@ contract BookMarketplace is ERC721URIStorage {
     }
 
     /* get the listing for a given token id */
-    function getListingByTokenId(uint256 tokenId) public view returns (Listing memory) {
+    function getListingByTokenId(
+        uint256 tokenId
+    ) public view returns (Listing memory) {
         return idToMarketItem[tokenId];
     }
 
+    /* Creates the sale of a marketplace item */
+    /* Transfers ownership of the item, as well as funds between parties */
+    function createMarketSale(uint256 tokenId) public payable {
+        uint price = idToMarketItem[tokenId].instantPrice;
+        address seller = idToMarketItem[tokenId].seller;
+        require(
+            msg.value == price,
+            "Please submit the asking price in order to complete the purchase"
+        );
+        idToMarketItem[tokenId].owner = payable(msg.sender);
+        idToMarketItem[tokenId].sold = true;
+        idToMarketItem[tokenId].seller = payable(address(0));
+        _itemsSold.increment();
+        _transfer(address(this), msg.sender, tokenId);
+        payable(marketOwner).transfer(listingPrice);
+        payable(seller).transfer(msg.value);
+    }
+
+    /* Creates the sale of a marketplace item */
+    /* Transfers ownership of the item, as well as funds between parties */
+    function createBiddingSale(
+        uint256 tokenId
+    ) public onlyListingSeller(tokenId) {
+        require(
+            idToMarketItem[tokenId].allowBid,
+            "This listing does not allow bidding"
+        );
+
+        int256 highestBidIndex = getHighestBidIndex(tokenId);
+        require(highestBidIndex > 0, "This listing does not have any bid yet");
```

Comment:
As there is an empty bid initialized in the list, will this ever happen?

Reply:
that empty bid will always have an index of 0, and as highestBidIndex must be > 0, the requirement will only pass if there is an "actual" bid.

Reply:
Oh, duh. My brain also seems to be toast today haha.